PUSH_TO_FILE.md

```diff
@@ -82,201 +292,220 @@ for a description of each environment variable setting:
+
+For a full list of global Go text template functions, reference the official
+[`text/template` documentation](https://pkg.go.dev/text/template#hdr-Functions).
+
+### Execution "Double-Pass"
+
+To avoid leaking sensitive secret data to logs, and to ensure that a
+misconfigured Push to File workflow fails fast, Push to File implements a
+"double-pass" execution of custom templates. The template "first-pass" runs
+before secrets are retrieved from Conjur, and validates that the provided custom
+template successfully executes given `"REDACTED"` for each secret value.
+Redacting secret values allows for secure, complete error logging for
+malformed templates. The template "second-pass" runs when rendering secret
+files, and error messages during this stage are sanitized. Custom templates that
+pass the "first-pass" and fail the "second-pass" require experimenting locally
+to identify bugs.
+
+_**NOTE**: Custom templates should not branch conditionally on secret values.
+This may result in a template first-pass execution that doesn't validate all
+branches of the custom template._
+
+### Example Custom Templates: Direct reference to secret values
+
+The following is an example of using a custom template to render secret data by
+referencing secrets directly using the custom template function `secret`.
 
-This will create a file /opt/secrets/conjur/redis.yaml, with contents as below.
 ```
-"api-url": "value-dev/redis/api-url"
-"admin-username": "value-dev/redis/username"
-"admin-password": "value-dev/redis/password"
+conjur.org/secret-file-template.direct-reference: |
+  username | {{ secret "db-username" }}
+  password | {{ secret "db-password" }}
 ```
 
-Below are code snippet is for JSON output.
+Assuming that the following secrets have been retrieved for secret group
+`direct-reference`:
 
 ```
-conjur.org/conjur-secrets.cache: |
-  - dev/redis/api-url
-  - admin-username: dev/redis/username
-  - admin-password: dev/redis/password
-     conjur.org/secret-file-path.cache: "./testdata/redis.json"
-     conjur.org/secret-file-format.cache: "json"
+db-username: admin
+db-password: my$ecretP@ss!
 ```
 
-This will create a file redis.json, with contents as below.
+
 ```
-{"api-url":"value-dev/redis/api-url","admin-username":"value-dev/redis/username","admin-password
-":"value-dev/redis/password"}
+username | admin
+password | my$ecretP@ss!
 ```
 
-Below are code snippet is for Bash output.
+### Example Custom Templates: Iterative approach
+
+The following is an example of using a custom template to render secret data
+using an iterative process instead of referencing all variables directly.
 
 ```
-conjur.org/conjur-secrets.cache: |
-  - dev/redis/api-url
-  - admin-username: dev/redis/username
-  - admin-password: dev/redis/password
-     conjur.org/secret-file-path.cache: "./testdata/redis.sh"
-     conjur.org/secret-file-format.cache: "bash"
+conjur.org/secret-file-template.iterative-reference: |
+  {{- range $index, $secret := .SecretsArray -}}
```

Comment:
It might help users to have some background context for the `.SecretsArray` reference here. Maybe something like:

"_Here, the `.SecretsArray` is a reference to Secrets Provider's internal array of secrets that have been retrieved  from Conjur. For each entry in this array, there is a secret `Alias` and a secret `Value` field that can be referenced in the custom template._"

Reply:
Oh, good idea. Added!